Make FrogFS standalone builds self‑contained

illusionfield opened the pull request:
# Make FrogFS standalone builds self‑contained

optional Python deps via YAML and CMake (fix missing heatshrink2 in venv)

## Motivation

### Problem

The build venv did not include heatshrink2, so compress heatshrink rules failed unless we preinstalled it manually. In our CI we build a GUI binary without ESP‑IDF and deploy it automatically, so the FrogFS toolchain must work entirely standalone.

### Goal

Allow projects to declare extra Python packages/requirements that are installed into the FrogFS build virtualenv so optional compressors (like heatshrink) “just work” without global pip state.

## How We Use FrogFS Today

- Build the web app, collect artifacts into a resources dir, and compress almost everything with heatshrink:

```yaml
frogfs.yaml declares
collect:
  - $cwd/resources/
filter:
  '*':
    - compress heatshrink
```

- We add

```yaml
python_packages:
  - heatshrink2>=0.13.0
```

so the compressor is available at build time.

- Minimal CMake driving only the FrogFS binary generation (no ESP‑IDF):

```cmake
include(frogfs/cmake/functions.cmake)
declare_frogfs_bin(NAME gui)
```

- CI step builds `generate_gui_bin` and publishes `build/CMakeFiles/gui.bin`.

- Our deploy system consumes `gui.bin` directly, so FrogFS must bootstrap its own Python deps in its venv for a hermetic, automated pipeline.

## What Changed

### tools

- `tools/mkfrogfs.py`: installs extra Python deps declared in `frogfs.yaml` into the FrogFS build venv, with caching.
- Adds config keys: `python_requirements` and `python_packages` (also supported under a nested `python:` block as `python.requirements` and `python.packages`).
- Content‑hash caching to avoid redundant installs; stored as `frogfs-venv-extra.stamp`.
- Re‑attempts importing `heatshrink2` after installs.

### CMake

- `cmake/functions.cmake`: expose two new args to preinstall extras into the venv:
  - `REQUIREMENTS <path>`: additional requirements.txt.
  - `PIP "<specs...>"`: space‑separated requirement specifiers.

## Why This Solves the Issue

Declaring

```yaml
python_packages: [heatshrink2>=0.13.0]
```

in `frogfs.yaml` ensures `mkfrogfs.py` installs the compressor into the FrogFS venv before preprocessing. No more reliance on global environment or ESP‑IDF.

The same is achievable from CMake using `REQUIREMENTS` and/or `PIP` for projects that prefer not to modify `frogfs.yaml`.

Builds remain reproducible and hermetic; caching avoids repeated installs.

## Usage Examples

### YAML (project‑local, CI‑friendly)

```yaml
frogfs.yaml:
python_packages:
  - heatshrink2>=0.13.0
filter:
  '*':
    - compress heatshrink
```

### CMake (centralized control)

```cmake
target_add_frogfs(... REQUIREMENTS ${CMAKE_SOURCE_DIR}/tools/frogfs-extra-requirements.txt PIP "heatshrink2>=0.13.0")
```

### Standalone build

Declare

```cmake
declare_frogfs_bin(NAME gui)
```

then

```bash
cmake --build <build> --target <target>
```

artifact at `<build>/CMakeFiles/<NAME>.bin`.

illusionfield committed:
Extra Python dependencies

illusionfield committed:
bugfix

jkent commented on README.md:
```diff
@@ -137,6 +150,47 @@ Verbs are applied in descending order. You can prefix a transforms or the
+- `python_packages`: a list of requirement specifiers (e.g. `pkg`, `pkg==1.2`).
+
+The same options are also supported under a nested `python:` section as
+`python.requirements` and `python.packages`.
```
I'm not sure if I like the two different ways of doing things.  In the spirit of the way the YAML is currently implemented, lets go with the toplevel python dict with requirements and packages keys.  Otherwise I really like this.

illusionfield replied:
Sounds great, thanks! In the draft PR I added both forms just because I wasn’t sure about the repo’s conventions, but this makes it clear. I’ll update it to only use the top-level keys. Appreciate the feedback!

jkent reviewed (changes requested):
Thank you for this fabulous PR.  I've only got the one issue that I'd like you to address.

illusionfield committed:
Remove support for nested `python:` section and keep only
toplevel `python_requirements` and `python_packages` keys,
to align with existing YAML conventions.

jkent reviewed (approved):
👍
